server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java

```diff
@@ -0,0 +1,161 @@
+        @SuppressWarnings("unchecked")
+        List<Float> vector = (List<Float>) args[1];
+        float[] vectorArray = new float[vector.size()];
+        for (int i = 0; i < vector.size(); i++) {
+            vectorArray[i] = vector.get(i);
+        }
+        return new KnnSearchBuilder((String) args[0], vectorArray, (int) args[2], (int) args[3]);
+    });
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
+        PARSER.declareFloatArray(constructorArg(), QUERY_VECTOR_FIELD);
+        PARSER.declareInt(constructorArg(), K_FIELD);
+        PARSER.declareInt(constructorArg(), NUM_CANDS_FIELD);
+        PARSER.declareFloat(KnnSearchBuilder::boost, BOOST_FIELD);
+    }
+
+    public static KnnSearchBuilder fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    // visible for testing
+    final String field;
+    final float[] queryVector;
+    final int k;
+    final int numCands;
+    float boost = AbstractQueryBuilder.DEFAULT_BOOST;
+
+    /**
+     * Defines a kNN search.
+     *
+     * @param field       the name of the vector field to search against
+     * @param queryVector the query vector
+     * @param k           the final number of nearest neighbors to return as top hits
+     * @param numCands    the number of nearest neighbor candidates to consider per shard
+     */
+    public KnnSearchBuilder(String field, float[] queryVector, int k, int numCands) {
+        this.field = field;
+        this.queryVector = queryVector;
+        this.k = k;
+        this.numCands = numCands;
+    }
+
+    public KnnSearchBuilder(StreamInput in) throws IOException {
+        this.field = in.readString();
+        this.k = in.readVInt();
+        this.numCands = in.readVInt();
+        this.queryVector = in.readFloatArray();
+        this.boost = in.readFloat();
+    }
+
+    /**
+     * An optional boost to apply to the kNN search scores.
+     */
+    public KnnSearchBuilder boost(float boost) {
+        this.boost = boost;
+        return this;
+    }
+
+    /**
+     * The number of nearest neighbors to return as top hits.
+     */
+    public int k() {
+        return k;
+    }
+
+    public KnnVectorQueryBuilder toQueryBuilder() {
+        if (k < 1) {
+            throw new IllegalArgumentException("[" + K_FIELD.getPreferredName() + "] must be greater than 0");
+        }
+        if (numCands < k) {
+            throw new IllegalArgumentException(
+                "[" + NUM_CANDS_FIELD.getPreferredName() + "] cannot be less than " + "[" + K_FIELD.getPreferredName() + "]"
+            );
+        }
+        if (numCands > NUM_CANDS_LIMIT) {
+            throw new IllegalArgumentException("[" + NUM_CANDS_FIELD.getPreferredName() + "] cannot exceed [" + NUM_CANDS_LIMIT + "]");
+        }
+        return new KnnVectorQueryBuilder(field, queryVector, numCands).boost(boost);
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
Should `equal` and `hashCode` also include `boost` field?

Reply:
Oops, will fix this.